src/core/interpreter.cc

```diff
@@ -892,17 +894,38 @@ Interpreter* InterpreterManager::Get() {
 }
 
 void InterpreterManager::Return(Interpreter* ir) {
-  DCHECK_LE(storage_.data(), ir);                    // ensure the pointer
-  DCHECK_GE(storage_.data() + storage_.size(), ir);  // belongs to storage_
-  available_.push_back(ir);
-  waker_.notify();
+  if (ir >= storage_.data() && ir < storage_.data() + storage_.size()) {
+    available_.push_back(ir);
+    waker_.notify();
+  } else if (return_untracked_ > 0) {
+    return_untracked_--;
+    if (return_untracked_ == 0) {
+      reset_ec_.notify();
+    }
+  } else {
+    LOG(DFATAL) << "Returning untracked interpreter";
+  }
 }
 
 void InterpreterManager::Reset() {
-  waker_.await([this]() { return available_.size() == storage_.size(); });
+  lock_guard guard{reset_mu_};
+
+  // we perform double buffer swapping with storage and wait for the old interepreters to be
+  // returned.
+  return_untracked_ = storage_.size() - available_.size();
+
+  std::vector<Interpreter> next_storage;
+  next_storage.reserve(storage_.capacity());
+  next_storage.resize(storage_.size());
```

Comment:
We probably don't need to create many interpreters at once

Reply:
what do you suggest? removing the resize?

Reply:
Yes, but that's not necessarily, it just spreads the latency